app/product_notices.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/mattermost/mattermost-server/v5/config"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/utils"
+	"github.com/pkg/errors"
+	"github.com/reflog/dateconstraints"
+)
+
+const MAX_REPEAT_VIEWINGS = 3
+
+// where to fetch notices from. setting as var to allow overriding during build/test
+var NOTICES_JSON_URL = "https://raw.githubusercontent.com/reflog/notices-experiment/master/notices.json"
+
+// http request cache
+var noticesCache = utils.RequestCache{}
+
+// cached counts that are used during notice condition validation
+var cachedPostCount int64
+var cachedUserCount int64
+
+// previously fetched notices
+var cachedNotices model.ProductNotices
+
+func noticeMatchesConditions(a *App, client model.NoticeClientType, clientVersion, locale string, postCount, userCount int64, isSystemAdmin, isTeamAdmin bool, isCloud bool, sku string, notice *model.ProductNotice) (bool, error) {
```

Comment:
Looks like you are receiving an `*App` parameter but you are not using it.

Reply:
@jespino *App is being used by validateConfigEntry function

Reply:
o yes, I didn't see that, but probably you only need to pass the current config, right? not the whole App

Reply:
yup. i'll fix that for clarity